Solutions/Menes.Testing.AspNetCoreSelfHosting/Menes/Testing/AspNetCoreSelfHosting/Internal/OpenApiWebHostStartup.cs

```diff
@@ -0,0 +1,73 @@
+{
+    using System;
+    using Menes;
+    using Menes.Hosting.AspNetCore;
+    using Microsoft.AspNetCore.Builder;
+    using Microsoft.AspNetCore.Hosting;
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Abstractions;
+    using Microsoft.AspNetCore.Routing;
+    using Microsoft.Azure.WebJobs.Hosting;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Startup class used with <see cref="IWebHostBuilder"/> to initialise a webhost using an <see cref="IWebJobsStartup"/>
+    /// implementation from a functions app.
+    /// </summary>
+    /// <typeparam name="TWebJobStartup">
+    /// The type of the Startup class to use to configure the <see cref="IServiceCollection"/>.
+    /// </typeparam>
+    internal class OpenApiWebHostStartup<TWebJobStartup>
+        where TWebJobStartup : IWebJobsStartup, new()
+    {
+        /// <summary>
+        /// Configures the <see cref="IServiceCollection"/>, adding MVC routing and then handing off to the specified
+        /// Startup class to add further required services.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/> to add to.</param>
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddRouting();
+
+            var builder = new WebJobBuilder(services);
+            var targetStartup = new TWebJobStartup();
+            targetStartup.Configure(builder);
+        }
+
+        /// <summary>
+        /// Configures the function host, adding a catch-all route that then hands off to Menes to process the request.
+        /// </summary>
+        /// <param name="app">The <see cref="IApplicationBuilder"/> to configure.</param>
+        public void Configure(IApplicationBuilder app)
+        {
+            var openApiRouteHandler = new RouteHandler(
+                async context =>
+                {
+                    try
+                    {
+                        IOpenApiHost<HttpRequest, IActionResult> handler = context.RequestServices.GetRequiredService<IOpenApiHost<HttpRequest, IActionResult>>();
+                        IActionResult result = await handler.HandleRequestAsync(context.Request, context).ConfigureAwait(false);
+                        var actionContext = new ActionContext(context, context.GetRouteData(), new ActionDescriptor());
+                        await result.ExecuteResultAsync(actionContext).ConfigureAwait(false);
+                    }
+                    catch (Exception ex)
+                    {
+                        context.Response.StatusCode = 500;
+                        await context.Response.WriteAsync(ex.ToString()).ConfigureAwait(false);
+                    }
+                });
+
+            var routeBuilder = new RouteBuilder(app, openApiRouteHandler);
+
+            routeBuilder.MapRoute("CatchAll", "{*path}");
```

Comment:
Given this, could we not now just add the routes for the things we find in the OpenAPI document?

Reply:
Yes, but I don't want to do that unless Menes works in the same way - and our default approach for Menes-driven services is to allow it to handle all requests.